frontend/src/metabase/visualizer/visualizer.slice.ts

```diff
@@ -526,64 +511,34 @@ const visualizerSlice = createSlice({
             ...initialState,
           };
         }
-
-        maybeUpdateHistory(state, action);
+        Object.assign(state, initialState);
       })
-      .addCase(handleDrop, (state, action) => {
-        state.draggedItem = null;
-        maybeUpdateHistory(
-          state,
-          handleDropInner({
-            event: action.payload,
-            datasetMap: state.datasets,
-            dataSourceMap: Object.fromEntries(
-              state.cards.map((card) => {
-                const dataSource = createDataSource("card", card.id, card.name);
-                return [dataSource.id, dataSource];
-              }),
-            ),
-          }),
-        );
-      })
-      .addCase(removeDataSource, (state, action) => {
-        const source = action.payload;
-        if (source.type === "card") {
-          const cardId = source.sourceId;
-          state.cards = state.cards.filter((card) => card.id !== cardId);
-        }
-        delete state.expandedDataSources[source.id];
-        delete state.loadingDataSources[source.id];
-        delete state.datasets[source.id];
-        delete state.loadingDatasets[source.id];
-
-        const isLastDataSource = Object.keys(state.datasets).length === 0;
-        if (isLastDataSource) {
-          const present = _.clone(state.present);
-          state.past = [...state.past, present];
-          state.present = getInitialVisualizerHistoryItem();
-          state.future = [];
-          state.isVizSettingsSidebarOpen = false;
-          state.isSwapAffordanceVisible = false;
-        } else {
-          maybeUpdateHistory(state, action);
+      .addCase(addDataSource.fulfilled, (state, action) => {
+        const nextState = action.payload;
+        if (nextState) {
+          state.display = nextState.display;
+          state.columns = copy(nextState.columns);
+          state.columnValuesMapping = copy(nextState.columnValuesMapping);
+          state.settings = copy(nextState.settings);
         }
       })
       .addCase(fetchCard.pending, (state, action) => {
         const cardId = action.meta.arg;
         state.loadingDataSources[`card:${cardId}`] = true;
         state.error = null;
       })
-      .addCase(fetchCard.fulfilled, (state, action) => {
+      .addCase(fetchCard.fulfilled, (state, action: PayloadAction<Card>) => {
         const card = action.payload;
         const index = state.cards.findIndex((c) => c.id === card.id);
+
+        // `any` prevents the "Type instantiation is excessively deep" error
         if (index !== -1) {
-          state.cards[index] = card;
+          state.cards[index] = card as any;
         } else {
-          state.cards.push(card);
+          state.cards.push(card as any);
         }
+
         state.loadingDataSources[`card:${card.id}`] = false;
```

Comment:
Had to cast `card: Card` to `any` here to avoid the error. Would appreciate any ideas here, I just gave up lol